WooCommerce: Add product variation card component

justinshreve opened the pull request:
This adds a new variation card component and moves the current product variation types form to it, so we can start implementing other parts of the variations UI and keep them contained, instead of in the root product form component.

The logic in the card is carried over directly from the existing merged code. #13158 will handle updating this to save our types to the Redux store.

To Test:

* `make run`
* Go to http://calypso.localhost:3000/store/products/add/<your site url>
* Verify that you can see the variations card and use the toggle.

mtias commented:
How does this component look? It'd be good to have a screenshot, it may make sense to add it to shared calypso components as well.

justinshreve commented:
@mtias This is just moving around some code for our products add form which has a separate card for product variations data. It currently looks like this -- definitely not a shared component. I think we will have some components that will make good shared candidates, like price inputs.

<img width="710" alt="screen shot 2017-04-18 at 10 08 03 am" src="https://cloud.githubusercontent.com/assets/689165/25143401/17b39370-241f-11e7-9f52-b5d630ea73a8.png">

allendav commented on client/extensions/woocommerce/app/products/product-form-variation-card.js:
```diff
@@ -0,0 +1,68 @@
+		super( props );
+
+		this.state = {
+			isVariation: props.product && 'variable' === props.product.type ? true : false,
```
I'm occasionally confused by `isVariation` (I think of the "child" products when I casually read that) - maybe `isVariableProduct` might be clearer?

allendav replied:
Changed in https://github.com/Automattic/wp-calypso/pull/13155/commits/1d99a0df8199f86387e28865b5964e41993cdde3

allendav commented on client/extensions/woocommerce/app/products/product-form-variation-card.js:
```diff
@@ -0,0 +1,68 @@
+		this.handleToggle = this.handleToggle.bind( this );
+	}
+
+	handleToggle() {
```
Shouldn't this be using `editProduct` to change the product type?

allendav replied:
See also https://github.com/Automattic/wp-calypso/pull/13108#pullrequestreview-33639805 regarding having a helper for such things

allendav commented on client/extensions/woocommerce/app/products/product-form-variation-card.js:
```diff
@@ -0,0 +1,68 @@
+				expanded={ true }
+				className="products__variation-card"
+				header={ ( <FormToggle onChange={ this.handleToggle } checked={ this.state.isVariation }>
+				{variationToggleDescription}
```
Need white space around variable

allendav replied:
Changed in https://github.com/Automattic/wp-calypso/pull/13155/commits/1d99a0df8199f86387e28865b5964e41993cdde3

allendav commented on client/extensions/woocommerce/app/products/product-form.js:
```diff
@@ -2,15 +2,12 @@
 import ProductFormDetailsCard from './product-form-details-card';
-import ProductVariationTypesForm from './product-variation-types-form';
-import FormToggle from 'components/forms/form-toggle';
+import ProductFormVariationCard from './product-form-variation-card';
```
This is a nit, but would you consider having it be plural like the details card, i.e. `ProductFormVariationsCard`

allendav replied:
Changed in https://github.com/Automattic/wp-calypso/pull/13155/commits/1d99a0df8199f86387e28865b5964e41993cdde3

allendav commented on client/extensions/woocommerce/app/products/product-form-variation-card.js:
```diff
@@ -0,0 +1,68 @@
+				) }
+			>
+				{ this.state.isVariation && (
+					<ProductVariationTypesForm />
```
Will passing props (e.g. variations, actions, etc) to this component come in a later PR?

justinshreve commented:
Just a note that for some of these comments (passing props, using `editProduct`, and the question around state) are handled in #13158 since this PR is moving existing code around. However I'll look at the other notes here.

allendav commented:
> Just a note that for some of these comments (passing props, using editProduct, and the question around state) are handled in #13158 since this PR is moving existing code around. However I'll look at the other notes here.

OMG sorry - I got into reviewing this and forgot your opening comment about that above - my bad

allendav commented:
@coderkevin - i've made the tweaks i requested - mind reviewing anew?

coderkevin commented:
Although the card expands, the toggle isn't changing for me:
![variation_toggle](https://cloud.githubusercontent.com/assets/945228/25406625/f75a38e6-29cc-11e7-8ebd-c31068da4452.gif)

coderkevin commented on client/extensions/woocommerce/app/products/product-form-variations-card.js:
```diff
@@ -0,0 +1,68 @@
+				icon=""
+				expanded={ true }
+				className="products__variation-card"
+				header={ ( <FormToggle onChange={ this.handleToggle } checked={ this.state.isVariableProduct }>
```
@allendav, looks like you missed this one when you renamed `isVariation` to `isVariableProduct` which was causing the issue I mentioned in the previous comment.

I took the liberty of amending your commit with this single addition.

coderkevin commented:
Rebased.

DanReyLop commented on client/extensions/woocommerce/app/products/product-form.js:
```diff
@@ -2,15 +2,12 @@
-import FormToggle from 'components/forms/form-toggle';
+import ProductFormVariationsCard from './product-form-variations-card';
 
 export default class ProductForm extends Component {
```
Now that this component is stateless, it can be made functional. I don't know if you folks have a hard rule about that.

coderkevin replied:
Yeah, we should probably look into using a function for this if we can. @justinshreve and I will look it over to make sure it is won't gain any more state/complexity and if not, make it a function.

DanReyLop commented on client/extensions/woocommerce/app/products/product-form-variations-card.js:
```diff
@@ -0,0 +1,68 @@
+
+	handleToggle() {
+		this.setState( ( prevState ) => ( {
+			isVariableProduct: ! prevState.isVariableProduct,
```
Why is this only updated in the internal state of the component? Shouldn't it be dispatched to the Redux tree? I'm assuming there are still a few pieces missing and this is one of them, but I just want to double-check.

DanReyLop commented on client/extensions/woocommerce/app/products/product-form-variations-card.js:
```diff
@@ -0,0 +1,68 @@
+		const variationToggleDescription = i18n.translate(
+			'%(productName)s has variations, for example size and color.', {
+				args: {
+					productName: product && product.name || i18n.translate( 'This product' )
```
This structure can be confusing, I believe an example like this was regarded as "too clever" in an Slack conversation. It's very concise if you know JavaScript thoroughly including operator precedence, but I would honestly prefer a ternary here :)

coderkevin replied:
I don't feel this is much more confusing than a ternary, but I do agree one one thing: There should probably be parens around the first && operation at least, so the operator order is clearer.

DanReyLop commented on client/extensions/woocommerce/app/products/product-form-variations-card.js:
```diff
@@ -0,0 +1,68 @@
+
+	render() {
+		const { product } = this.props;
+		const variationToggleDescription = i18n.translate(
```
Is this the way to translate now? Shouldn't you use the `localize` HoC instead?

coderkevin replied:
It can be done either way, but you've got a fair point. I think new development is supposed to be going to using the HOC.

DanReyLop commented on client/extensions/woocommerce/app/products/product-form-variations-card.js:
```diff
@@ -0,0 +1,68 @@
+		return (
+			<FoldableCard
+				icon=""
+				expanded={ true }
```
This line can be simplified to `expanded`. In JSX, an attribute without value is assumed to be `={ true }`

coderkevin replied:
Thanks for the tip!

coderkevin commented:
Thanks for the great review, @DanReyLop! Either @justinshreve or I will have updates shortly.

justinshreve commented:
Updates:
* Fixed expanded attribute.
* Switched to using the localize HOC for translation.

I'm leaving the functional component changes for #13158 where the last of the internal state changes are removed.

I think this is ready for another look @DanReyLop @coderkevin @allendav.

coderkevin commented:
LGTM :shipit: 🚢
